compiler/src/dotty/tools/dotc/transform/IsInstanceOfChecker.scala

```diff
@@ -0,0 +1,133 @@
+package dotty.tools.dotc
+package transform
+
+import util.Positions._
+import MegaPhase.MiniPhase
+import core._
+import Contexts.Context, Types._, Decorators._, Symbols._, typer._, ast._, NameKinds._
+import TypeUtils._, Flags._
+import config.Printers.{ transforms => debug }
+
+/** Check runtime realizability of type test, see the documentation for `Checkable`.
+ */
+class IsInstanceOfChecker extends MiniPhase {
+
+  import ast.tpd._
+
+  val phaseName = "isInstanceOfChecker"
+
+  override def transformTypeApply(tree: TypeApply)(implicit ctx: Context): Tree = {
+    def ensureCheckable(qual: Tree, pt: Tree): Tree = {
+      if (!Checkable.checkable(qual.tpe, pt.tpe, tree.pos))
+         ctx.warning(
+           s"the type test for ${pt.show} cannot be checked at runtime",
+           tree.pos
+         )
+
+      tree
+    }
+
+    tree.fun match {
+      case fn: Select if fn.symbol == defn.Any_typeTest =>
+        ensureCheckable(fn.qualifier, tree.args.head)
+      case fn: Select if fn.symbol == defn.Any_isInstanceOf =>
+        ensureCheckable(fn.qualifier, tree.args.head)
+      case _ => tree
+    }
+  }
+}
+
+object Checkable {
+  import Inferencing._
+  import ProtoTypes._
+
+  /** Whether `(x:X).isInstanceOf[P]` can be checked at runtime?
+   *
+   *  First do the following substitution:
+   *  (a) replace `T @unchecked` and pattern binder types (e.g., `_$1`) in P with WildcardType
+   *  (b) replace pattern binder types (e.g., `_$1`) in X:
+   *      - variance = 1  : hiBound
+   *      - variance = -1 : loBound
+   *      - variance = 0  : OrType(Any, Nothing) // TODO: use original type param bounds
+   *
+   *  Then check:
+   *
+   *  1. if `X <:< P`, TRUE
+   *  2. if `P` is a singleton type, TRUE
+   *  3. if `P` refers to an abstract type member or type parameter, FALSE
+   *  4. if `P = Array[T]`, checkable(E, T) where `E` is the element type of `X`, defaults to `Any`.
```

Comment:
Can you explain why there is a special treatment for Array?

Reply:
Thanks @allanrenucci for the helpful feedback, I've addressed them in the latest commit.

For `Array`,  we need a special treatment, because JVM carries type information of array elements, it's possible to test the element type at runtime `[1]`:

```Scala
scala> Array(1, 2, 3).getClass
res2: Class[_ <: Array[Int]] = class [I
// where the string "[I" is the run-time type signature for the class 
// object "array with component type int".
```

`[1]`: https://docs.oracle.com/javase/specs/jls/se7/html/jls-10.html